### ISSUE TITLE
Guidance/support for evaluating on ScanNet++ (84-category taxonomy)?

### ISSUE THREAD

**Issue by ChenYutongTHU** [IDEA-Research/SegVGGT#2]
Thanks for releasing SegVGGT — I'd like to evaluate it on ScanNet++, which uses an 84-category taxonomy, different from the ScanNet (20) and ScanNet200 (200) setups currently documented.

From reading the code, it looks like the instance classification head is a closed-set linear layer sized directly from the `enable_semantic` config value:

```yaml
# configs/eval/segvggt_scannetv2.yaml
enable_semantic: 20
```

```python
# segvggt/heads/semantic_head.py
if enable_instance_seg:
    self.scratch.output_instance = nn.Sequential(
        nn.Linear(instance_cls_input_dim, instance_cls_hidden_dim),
        nn.ReLU(inplace=True),
        nn.Linear(instance_cls_hidden_dim, output_dim - 3)
    )
```

This produces exactly `num_classes + 1` logits (the `+1` being the "no object" class), so a checkpoint trained with `enable_semantic: 20` or `200` can't be pointed directly at ScanNet++'s 84-category ground truth — the output dimension and learned class-index semantics don't match.

A few questions:

1. Is there a released or planned config/checkpoint for ScanNet++ (`enable_semantic: 84`)?
2. If not, what's the recommended path to adapt an existing checkpoint to a new taxonomy — reinitialize just `output_instance` and fine-tune, or does the whole instance-seg branch (including query embeddings) need retraining from scratch?
3. Is the mask-prediction branch (the per-query binary segmentation logits) actually class-agnostic and reusable across taxonomies, or are the instance queries themselves implicitly tied to the trained category count in some way?

Any pointers would be appreciated.

**Comment by JinyuanQu** [comment]
Hi @ChenYutongTHU, thanks for your interest and for taking a close look at our code!

As discussed in Sec. A.7 of our paper, the current model uses a closed-set taxonomy, so the checkpoint’s learned class indices cannot directly represent the ScanNet++ categories. Accordingly, our ScanNet++ experiments in Sec. 4.3 follow a class-agnostic setting. We train the model only on ScanNet200 and evaluate it directly on ScanNet++ without fine-tuning to assess its cross-dataset generalization.

We have not trained on ScanNet++ or released a dedicated checkpoint for it. If you would like to explore this, please refer to the training discussion in [issue #1](https://github.com/IDEA-Research/SegVGGT/issues/1). We believe adapting a released checkpoint through fine-tuning should be feasible. A reasonable starting point would be to replace the classifier’s final linear layer with one that predicts 84 categories plus a no-object class, update the configuration and label mapping, and fine-tune using the pretrained queries and mask branch as initialization. We have not yet validated this setting experimentally.

Regarding question 3, the mask branch is class-agnostic in structure. Each query predicts a binary instance mask, and the query count is independent of the number of categories. Queries are not assigned to fixed semantic classes, so a new taxonomy does not require reinitializing them. However, since their features are learned jointly with classification supervision, they may retain biases from the training data and benefit from fine-tuning.

**Comment by wnn2000** [comment]
> Hi [@ChenYutongTHU](https://github.com/ChenYutongTHU), thanks for your interest and for taking a close look at our code!
> 
> As discussed in Sec. A.7 of our paper, the current model uses a closed-set taxonomy, so the checkpoint’s learned class indices cannot directly represent the ScanNet++ categories. Accordingly, our ScanNet++ experiments in Sec. 4.3 follow a class-agnostic setting. We train the model only on ScanNet200 and evaluate it directly on ScanNet++ without fine-tuning to assess its cross-dataset generalization.
> 
> We have not trained on ScanNet++ or released a dedicated checkpoint for it. If you would like to explore this, please refer to the training discussion in [issue #1](https://github.com/IDEA-Research/SegVGGT/issues/1). We believe adapting a released checkpoint through fine-tuning should be feasible. A reasonable starting point would be to replace the classifier’s final linear layer with one that predicts 84 categories plus a no-object class, update the configuration and label mapping, and fine-tune using the pretrained queries and mask branch as initialization. We have not yet validated this setting experimentally.
> 
> Regarding question 3, the mask branch is class-agnostic in structure. Each query predicts a binary instance mask, and the query count is independent of the number of categories. Queries are not assigned to fixed semantic classes, so a new taxonomy does not require reinitializing them. However, since their features are learned jointly with classification supervision, they may retain biases from the training data and benefit from fine-tuning.

When evaluating a model trained on ScanNet200 directly on ScanNet++, it may predict valid object instances that are not annotated in ScanNet++. These predictions would be counted as false positives, potentially leading to a low mAP.

How do you address this issue in your class-agnostic cross-dataset evaluation? Do you use any ignore-region handling, annotation filtering, or alternative metrics? More generally, do you think mAP is a reliable metric for this setting, given potentially incomplete instance annotations?

**Comment by JinyuanQu** [comment]
Hi @wnn2000, thanks for raising this. We agree that differences in annotation coverage can affect class-agnostic mAP. However, predictions of unannotated objects are not automatically counted as false positives in our evaluation.

We follow commonly used ScanNet-style instance evaluation rules. For ScanNet++, we retain GT instances belonging to its benchmark instance categories and mark the remaining points as ignored. An unmatched prediction is excluded from false-positive counting when the fraction of its points in ignored regions exceeds the current IoU threshold. In our ScanNet++ benchmark, all compared methods are evaluated using the same protocol, including the same GT category filtering and ignore rules, to ensure a fair comparison.

We evaluate the ScanNet200-trained model directly without fine-tuning or additional filtering based on its predicted semantic categories. We report class-agnostic mAP, AP50, and AP25 without introducing alternative metrics.

We consider mAP useful for comparing methods under this shared protocol. The ignore rules mitigate the annotation-coverage issue, but do not fully eliminate it. The scores therefore measure segmentation performance within the benchmark’s annotation scope, rather than the ability to segment all real objects in a scene.

**Comment by wnn2000** [comment]
@JinyuanQu 
Thank you for the clarification. I agree that introducing ignore regions makes this evaluation more appropriate. Since the model was trained with a richer category taxonomy than the test dataset annotates, a class-agnostic evaluation may produce valid object predictions that have no corresponding annotation in the test set. Without an ignore rule, these could be counted as false positives.

Is the ignore-region protocol you use a standard evaluation strategy adopted from prior work, or was it designed specifically for your benchmark? If it follows an existing protocol, I would appreciate a reference or implementation.

**Comment by JinyuanQu** [comment]
Hi @wnn2000, the ignore rule follows the [ScanNet instance evaluator](https://github.com/ScanNet/ScanNet/blob/master/BenchmarkScripts/3d_evaluation/evaluate_semantic_instance.py) and is also used by the [official ScanNet++ evaluator](https://github.com/scannetpp/scannetpp/blob/main/semantic/eval/eval_instance.py). To my knowledge, many prior methods, such as [OneFormer3D](https://github.com/filaPro/oneformer3d/blob/main/oneformer3d/evaluate_semantic_instance.py), [MAFT](https://github.com/JIA-Lab-research/Mask-Attention-Free-Transformer/blob/main/maft/evaluation/instance_eval.py), and [SegDINO3D](https://github.com/IDEA-Research/SegDINO3D/blob/main/evaluation/utils_instance_seg_3d_eval.py), use the same ignored-overlap rule.

Our [ScanNet200 evaluation](https://github.com/IDEA-Research/SegVGGT/blob/main/eval/instance_seg_eval.py) shares the same AP matching and false-positive rules. For ScanNet++, we retain its benchmark instance categories, map them to one foreground class for class-agnostic evaluation, and mark other GT points as void. An unmatched prediction is excluded from false positives when its ignored-point fraction exceeds the IoU threshold.
